### PR TITLE
Fix BaseModel type annotations to be resolvable by typing.get_type_hints

### LINKED ISSUE #7623: BaseModel method type annotations cause typing.get_type_hints to raise NameError
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

While using a project like [autodoc_pydantic](https://github.com/mansenfranzen/) and Sphinx, it uses [sphinx-autodoc-typehints](https://github.com/tox-dev/sphinx-autodoc-typehints) and ultimately the stdlib [typing.get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) method to inspect pydantic models, their methods, etc.  It seems that the type annotations that pydantic is advertising are incomplete/incorrect, causing `NameError` exceptions when calling `get_type_hints`, even natively from the Python REPL.

```pycon
>>> import typing
>>> from pydantic import BaseModel
>>> typing.get_type_hints(BaseModel.model_copy)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/typing.py", line 1871, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/usr/lib/python3.10/typing.py", line 327, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/usr/lib/python3.10/typing.py", line 694, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'Model' is not defined

>>> BaseModel.copy.__annotations__
{'self': 'Model', 'include': 'AbstractSetIntStr | MappingIntStrAny | None', 'exclude': 'AbstractSetIntStr | MappingIntStrAny | None', 'update': 'typing.Dict[str, Any] | None', 'deep': 'bool', 'return': 'Model'}

>>> typing.get_type_hints(BaseModel.dict)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/typing.py", line 1871, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/usr/lib/python3.10/typing.py", line 327, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/usr/lib/python3.10/typing.py", line 694, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'IncEx' is not defined

```

Seems like the annotations need to be fully-qualified or some other similar solution?  It may also be because of the conditional around `typing.TYPE_CHECKING` when defining type vars like [Model](https://github.com/pydantic/pydantic/blob/734d3f9362a9b5751a0c536ac6a5ae96a0a8da34/pydantic/main.py#L47), `IncEx`, `Literal` (from `typing_extensions`), etc.

For background, ideally need compatibility from Python 3.8-3.11+.

### Example Code

```Python
import typing
from pydantic import BaseModel
typing.get_type_hints(BaseModel.model_copy)
typing.get_type_hints(BaseModel.model_dump)
typing.get_type_hints(BaseModel.dict)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.3.0
        pydantic-core version: 2.6.3
          pydantic-core build: profile=release pgo=true
                 install path: ....../env/lib/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]
                     platform: Linux-6.2.0-33-generic-x86_64-with-glibc2.35
     optional deps. installed: ['email-validator', 'typing-extensions']
```

### PR DESCRIPTION
**Pull request by devmonkey22**
Fix BaseModel type annotations to be resolvable by typing.get_type_hints (#7623)

## Change Summary

Updated required `main.py` type definitions used with non-deprecated `BaseModel` methods to always be defined, instead of conditionally when `typing.TYPE_CHECKING`.

## Related issue number

Fix #7623 

## Checklist

* [X] The pull request title is a good summary of the changes - it will be used in the changelog
* [X] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @davidhewitt

### CODE REVIEW

**Comment by sydney-runkle** [comment]
@devmonkey22,

Thanks for your contribution here. Is this ready to review? If so, could you please rebase to resolve the conflicts with `main`?

**Comment by sydney-runkle** [comment]
Currently running into some test issues related to this: https://github.com/pydantic/pydantic/issues/8307

**Comment by sydney-runkle** [comment]
I think the right path forward here is to implement a solution that works for Python 3.10+ and add xfails to tests for 3.8 and 3.9.

File: tests/test_type_hints.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    class ParentModel(BaseModel):
+        grape: bool
+        banana: UltraSimpleModel
+
+    return ParentModel
+
+
+def inspect_type_hints(
+    obj_type, members: Optional[Set[str]] = None, exclude_members: Optional[Set[str]] = None, recursion_limit: int = 3
+):
```
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply tests/test_type_hints.py:61]: Good question. They're pretty gnarly, so I'd opt to skip returning for now. Down the line, if someone's feeling quite invested, they could make the addition, but I don't see it as a blocker on this PR.

**Review by alexmojaki — COMMENTED** [review]
welp, i was in the process of reviewing

**Pull request merged**